gpjax/sparse_gps.py

```diff
@@ -0,0 +1,213 @@
+        
+        if not self.q.whiten:
+            z = params["q"]["inducing_inputs"]
+            mz = self.prior.mean_function(z, params["mean_function"])
+            Kzz = gram(self.prior.kernel, z, params["kernel"])
+            Kzz += I(nz) * self.jitter
+            Lz = cholesky(Kzz, lower=True)
+            pu = distrax.MultivariateNormalTri(mz.squeeze(), Lz)
+            
+        else:
+            pu = distrax.MultivariateNormalDiag(jnp.zeros(nz))
+        
+        return qu.kl_divergence(pu)
+
+
+    # Compute expectation, ∫ log(p(y|F)) q(F) dF, through Gauss-Hermite quadrature:
+    def variational_expectation(self, params: Dict, batch: Dataset) -> Array:
+        x, y = batch.X, batch.y
+
+        Fmu, Fvar = vmap(self.pred_moments, in_axes=(None, 0))(params, x[:,jnp.newaxis,:])
+
+        # Get log(p(y|F)) function for current likelihood parameter values:
+        def log_prob(F, y):
+            return self.likelihood.link_function(F, params["likelihood"]).log_prob(y)
+
+        return gauss_hermite_quadrature(log_prob, Fmu.squeeze(1), Fvar.squeeze(1), y=y)
+    
+    # Computes predictive moments for Gauss-Hermite quadrature:
+    def pred_moments(self, params: Dict, test_inputs: Array) -> Array:
+        mu = params["q"]["mu"]
+        sqrt = params["q"]["sqrt"]
+        
+        # Cholesky decomposition at inducing inputs:
+        z = params["q"]["inducing_inputs"]
+        nz = self.num_inducing
+        Kzz = gram(self.prior.kernel, z, params["kernel"])
+        Kzz += I(nz) * self.jitter
+        Lz = cholesky(Kzz)
+        
+        # Compute predictive moments:
+        t = test_inputs
+        Ktt = gram(self.prior.kernel, t, params["kernel"])
+        Kzt = cross_covariance(self.prior.kernel, z, t, params["kernel"])
+        M = solve_triangular(Lz, Kzt, lower=True)
+        Fcov = Ktt - jnp.matmul(M.T, M)
+    
+        if not self.q.whiten:
+            M = solve_triangular(Lz.T, M, lower=False)
+            mz = self.prior.mean(params)(z)
+            mu -= mz
+
+        Fmu = self.prior.mean(params)(t) + jnp.matmul(M.T, mu)
+        V = jnp.matmul(M.T, sqrt)
+        Fcov += jnp.matmul(V, V.T)
+        
+        return Fmu, Fcov
+
+    def mean(self, params: Dict) -> Callable[[Array], Array]:
+        # Cholesky decomposition at inducing inputs:
+        z = params["q"]["inducing_inputs"]
+        nz = self.num_inducing
+        Kzz = gram(self.prior.kernel, z, params["kernel"])
+        Kzz += I(nz) * self.jitter
+        Lz = cholesky(Kzz)
+
+        # Variational mean:
+        mu = params["q"]["mu"]
+        if not self.q.whiten:
+            mz = self.prior.mean(params)(z)
+            mu -= mz
+        
+        def mean_fn(test_inputs: Array):
+            t = test_inputs
+            Kzt = cross_covariance(self.prior.kernel, z, t, params["kernel"])
+            M = solve_triangular(Lz, Kzt, lower=True)
+    
+            if not self.q.whiten:
+                M = solve_triangular(Lz.T, M, lower=False)
```

Comment:
This conditional might be more cleanly handled as part of the variational family object.

Reply:
As above